### PR TITLE
Optimize suppress operations during formatting.  

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Drops us from:

image

To:

image

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ListExtensions.cs [hunk 1]
```diff
@@ -58,6 +59,25 @@ internal static class ListExtensions
         list.RemoveRange(targetIndex, list.Count - targetIndex);
     }
 
+    public static void RemoveOrTransformAll<T, TArg>(this ArrayBuilder<T> list, Func<T, TArg, T?> transform, TArg arg)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ListExtensions.cs:62]: copy of the method above.  just for ArrayBuilders.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Context/FormattingContext.InitialContextFinder.cs [hunk 2]
```diff
@@ -53,7 +55,6 @@ private class InitialContextFinder
                 if (initialSuppressOperations != null)
                 {
                     Debug.Assert(
-                        initialSuppressOperations.IsEmpty() ||
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Context/FormattingContext.InitialContextFinder.cs:56]: redundant with line below.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractFormatEngine.cs [hunk 3]
```diff
@@ -176,6 +176,20 @@ private static void AddOperations<T>(SegmentedList<T> operations, List<T> scratc
         scratch.Clear();
     }
 
+    private static void AddOperations<T>(SegmentedList<T> operations, ArrayBuilder<T> scratch, SyntaxNode node, Action<ArrayBuilder<T>, SyntaxNode> addOperations)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractFormatEngine.cs:179]: copy of method above, just for array builders.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Rules/Operations/FormattingOperations.cs [hunk 4]
```diff
@@ -141,18 +141,6 @@ public static AdjustSpacesOperation CreateAdjustSpacesOperation(int space, Adjus
...
-    /// return SuppressOperation for the node provided by the given formatting rules
-    /// </summary>
-    internal static IEnumerable<SuppressOperation> GetSuppressOperations(IEnumerable<AbstractFormattingRule> formattingRules, SyntaxNode node, SyntaxFormattingOptions options)
-    {
-        var chainedFormattingRules = new ChainedFormattingRules(formattingRules, options);
-
-        var list = new List<SuppressOperation>();
-        chainedFormattingRules.AddSuppressOperations(list, node);
-        return list;
-    }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Rules/Operations/FormattingOperations.cs:154]: not called at all.

File: src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService+CompatAbstractMetadataFormattingRule.cs [hunk 5]
```diff
@@ -18,7 +19,7 @@ protected abstract class CompatAbstractMetadataFormattingRule : AbstractMetadata
 #pragma warning disable CS0809 // Obsolete member overrides non-obsolete member
         [Obsolete("Do not call this method directly (it will Stack Overflow).", error: true)]
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public sealed override void AddSuppressOperations(List<SuppressOperation> list, SyntaxNode node, in NextSuppressOperationAction nextOperation)
+        public sealed override void AddSuppressOperations(ArrayBuilder<SuppressOperation> list, SyntaxNode node, in NextSuppressOperationAction nextOperation)
```
Comment (ToddGrun) [inline src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService+CompatAbstractMetadataFormattingRule.cs:22]: >ArrayBuilder [](http://example.com/codeflow?start=58&length=12)

Not worth doing for the other Add* methods in this file?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService+CompatAbstractMetadataFormattingRule.cs:22]: maybe one more based on traces.  but i'd like to do it one at a time, and based on need.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractFormatEngine.cs [hunk 6]
```diff
@@ -150,14 +150,14 @@ protected virtual NodeOperations CreateNodeOperations(CancellationToken cancella
...
-            AddOperations(nodeOperations.AlignmentOperation, alignmentOperation, node, addAlignTokensOperations);
-            AddOperations(nodeOperations.AnchorIndentationOperations, anchorIndentationOperations, node, addAnchorIndentationOperations);
+            AddOperations(nodeOperations.IndentBlockOperation, indentBlockOperationScratch, node, addIndentBlockOperations);
+            AddOperations(nodeOperations.SuppressOperation, suppressOperationScratch, node, addSuppressOperation);
+            AddOperations(nodeOperations.AlignmentOperation, alignmentOperationScratch, node, addAlignTokensOperations);
+            AddOperations(nodeOperations.AnchorIndentationOperations, anchorIndentationOperationsScratch, node, addAnchorIndentationOperations);
         }
 
         // make sure we order align operation from left to right
-        alignmentOperation.Sort(static (o1, o2) => o1.BaseToken.Span.CompareTo(o2.BaseToken.Span));
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractFormatEngine.cs:160]: >alignmentOperation [](http://example.com/codeflow?start=8&length=18)

was this a bug before?

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**